### PR TITLE
Don't forget the datatype when decoding a PEM file

### LINKED ISSUE #13046: Unable to load X9.42 DH Parameters using PEM_read_bio_DHparams (3.0.0-master)
Fails on master (compiled today), passes on 1.1.1h. I haven't checked to see if this is failing in older alphas.

```c
#include <stdio.h>
#include <string.h>
#include <openssl/bio.h>
#include <openssl/dh.h>
#include <openssl/pem.h>

int main() {
    char *params = "-----BEGIN X9.42 DH PARAMETERS-----\nMIICKQKCAQEArRB+HpEjqdDWYPqnlVnFH6INZOVoO5/RtUsVl7YdCnXm+hQd+VpW\n26+aPEB7od8V6z1oijCcGA4d5rhaEnSgpm0/gVKtasISkDfJ7e/aTfjZHo/vVbc5\nS3rVt9C2wSIHyfmNEe002/bGugssi7wnvmoA4KC5xJcIs7+KMXCRiDaBKGEwvImF\n2xYC5xRBXZMwJ4Jzx94x79xzEPcSH9WgdBWYfZrcCkhtzfk6zEQyg4cxXXXhmMZB\npIDNhqG55YfovmDmnMkosrnFIXLkEwQumyPxCw4W55djybU9z0uoCinj+3PBa451\nuX7zY+L/ox9xz53lOE5xuBwKxN/+DBDmTwKCAQEArEAy708tmuOd8wtcj/2sUGze\nvnuJmYyvdIZqCM/k/+OmgkpOELmm8N2SHwGnDEr6q3OddwDCn1LFfbF8YgqGUr5e\nkAGo1mrXwXZpEBmZAkr00CcnWsE0i7inYtBSG8mK4kcVBCLqHtQJk51U2nRgzbX2\nxrJQcXy+8YDrNBGOmNEZUppF1vg0Vm4wJeMWozDvu3eobwwasVsFGuPUKMj4rLcK\ngTcVC47rEOGD7dGZY93Z4mPkdwWJ72qiHn9fL/OBtTnM40CdE81Wavu0jWwBkYHh\nvP6UswJp7f5y/ptqpL17Wg8ccc//TBnEGOH27AF5gbwIfypwZbOEuJDTGR8r+gId\nAIAcDTTFjZP+mXF3EB+AU1pHOM68vziambNjces=\n-----END X9.42 DH PARAMETERS-----\n";
    BIO *bio = BIO_new_mem_buf(params, strlen(params));
    DH *dh = PEM_read_bio_DHparams(bio, NULL, NULL, NULL);
    if (dh == NULL) {
        printf("failed to load\n");
        return 1;
    }
    return 0;
}
```

### PR DESCRIPTION
**Pull request by mattcaswell**
The OSSL_STORE code was forgetting the datatype that we read from the
PEM header when decoding the DER.

Fixes #13046

WIP because it needs a test.

@levitte - I'm not sure where to put a test for this? Can you give me a hint?

### CODE REVIEW

**Comment by levitte** [comment]
> @levitte - I'm not sure where to put a test for this? Can you give me a hint?

I would write a separate recipe for this, that generates a X9.42 and a DSA PEM file, and capture the output of `openssl storeutl` on them, and see that you get the BEGIN line you expect for each.

**Comment by mattcaswell** [comment]
I've rebased this to fix a conflict with master and extended it with additional tests.

My first attempt at adding tests was to create a test that called various PEM_read_bio_* functions to check that we got the expected result back. This effectively does the same thing as the reproducer in #13046. To my surprise I found that the tests I had written were passing in master even without my fix applied!! On further investigation I found that issue in #13046 no longer exists. It was "fixed" by commit 35426b2f9c. However the "fix" only addresses the symptoms by changing `PEM_read_bio_DHparams()` to not use OSSL_STORE anymore. The bug in OSSL_STORE remains.

Therefore I have kept the original tests that I wrote (they would have caught the issue before 35426b2f9c was applied). I've also added more tests to test OSSL_STORE directly.

This PR is now out of WIP. Please take a look.

**Comment by mattcaswell** [comment]
Fixup pushed to address some missing no-dh and no-dsa guards.

File: test/pem_read_depr_test.c [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+ * warnings.
+ */
+#define OPENSSL_SUPPRESS_DEPRECATED
+
+#include <openssl/pem.h>
+#include <openssl/bio.h>
+
+#include "testutil.h"
+
+const char *datadir;
```
Comment (levitte) [inline test/pem_read_depr_test.c:21]: Add `static`

Comment (mattcaswell) [reply test/pem_read_depr_test.c:21]: Done

**Comment by levitte** [comment]
I saw a surprising memleak in the last Travis results...  have you tried an ASAN build?

**Comment by mattcaswell** [comment]
> I saw a surprising memleak in the last Travis results... have you tried an ASAN build?

Leak was in the test. Now fixed.

File: test/ossl_store_test.c [hunk 2]
```diff
@@ -76,9 +77,10 @@ static int get_params(const char *uri, const char *type)
     OSSL_STORE_close(ctx);
 
     if (pkey != NULL)
-        return EVP_PKEY_is_a(pkey, type);
+        ret = EVP_PKEY_is_a(pkey, type);
+    EVP_PKEY_free(pkey);
 
-    return 0;
+    return ret;
```
Comment (levitte) [inline test/ossl_store_test.c:83]: The `err` section doesn't look very "errory" any more.  How about changing all from `err:` to this?

``` C
    if (pkey != NULL)
        ret = EVP_PKEY_is_a(pkey, type);
    EVP_PKEY_free(pkey);

 err:
    OSSL_STORE_close(ctx);
    return ret;
```

**Review by levitte — CHANGES_REQUESTED** [review]

File: test/ossl_store_test.c (hunk 2 above)
Comment (mattcaswell) [reply test/ossl_store_test.c:83]: Done

**Review by levitte — APPROVED** [review]
If CIs agree...

**Comment by mattcaswell** [comment]
Travis failures are unrelated to this PR.

**Comment by levitte** [comment]
> Travis failures are unrelated to this PR.

Confirmed

**Comment by slontis** [comment]
@mattcaswell are you saying that this PR doesnt fix the ossl store issue? If so is there an issue for this?

**Comment by mattcaswell** [comment]
>  @mattcaswell are you saying that this PR doesnt fix the ossl store issue? If so is there an issue for this?

I assume you are referring to this comment by me:

>My first attempt at adding tests was to create a test that called various PEM_read_bio_* functions to check that we got the expected result back. This effectively does the same thing as the reproducer in #13046. To my surprise I found that the tests I had written were passing in master even without my fix applied!! On further investigation I found that issue in #13046 no longer exists. It was "fixed" by commit 35426b2. However the "fix" only addresses the symptoms by changing PEM_read_bio_DHparams() to not use OSSL_STORE anymore. The bug in OSSL_STORE remains.

What I was trying to say (apparently unsuccessfully) is that there has been a "fix" already applied that meant the reproducer in #13046 no longer works. But that other "fix" did not address the OSSL_STORE bug. *This* PR *does* address that bug.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**